### ISSUE TITLE
RTK Query - when invalidating, data should not become undefined, it should be accessible for better experience

### ISSUE THREAD

**Issue by omerman** [reduxjs/redux-toolkit#1409]
First of all thank you for your amazing work & tools.

Now to the issue in hand, ;)

When I invalidate an entry, I expect that the query holding that entry will still hold the invalidated data. or at least be able to do something like this:

Option A
```
/* data will also hold the invalidated data as well, meaning after invalidating, 
instead of data being undefined, it will hold the invalidated data until new data arrives. 
The indicator of isFetching should be true in that case, or make a new one, isValidating.
*/
const {data} = useMyQuery();
```

Option B
```
const {invalidatedData, data = invalidatedData} = useMyQuery();
```

Hope this is making sense to you as well & Looking forward hearing from you & Thanks again for your work :)

**Comment by markerikson** [comment]
As far as I know this is exactly the case. When you invalidate a query, the existing fetched data should stay available as `data` until the new request returns a result.  

Can you show an example of what you're seeing and describe what's different than what you expect to see?  A CodeSandbox would be helpful here.

**Comment by omerman** [comment]
@markerikson  Thank you for the quick answer! I'll make sure to add a CodeSandbox. Probably tomorrow.

**Comment by omerman** [comment]
@markerikson 

Here is a sandbox - https://codesandbox.io/s/stupefied-dream-2uxyi?file=/src/App.tsx

I added some instructions inside it that explains the issue.
Let me know if it's ok and you see the problem too.

Thanks again! :)

**Comment by phryneas** [comment]
You are observing something different there:

When something is invalidated, that means one of two things:

* the data is currently being used by a component => refetch it
* the data is currently not being used by a component => remove it from cache

So in your case, the data is simply being removed from the cache. And then your old component mounts again and fetches it from the server normally.  

We do this as you could have 100 cache entries that you will never display again and in that case we don't want to send 100 requests to the server.

**Comment by omerman** [comment]
@phryneas I  don't understand why you would treat that data differently. i think the best solution would be to treat the second case like the first one, and keep leveraging the `keepUnusedDataFor`. If the time expires, remove the invalidated data from the cache. If not, keep it until it is used / the right amount of time passes.

**Comment by phryneas** [comment]
Because we *do* the invalidation-induced refetch right away. 

We are not keeping semi-invalid data floating around. Imagine it gets changed to something completely different and you use it again 30 seconds later. You will have data flashing up that is totally wrong, at a time the user really does not expect that to be there any more. A second later? Sure, but 30 seconds later seeing something that should be deleted or something similar? Really bad UX.

If there is data you want to stay in the cache *and* to be refetched, then keep some component remounted that accesses it. You don't even need to display it.

**Comment by omerman** [comment]
@phryneas I understand what u are saying, but can you really tell the difference between this situation when there is a component using the data? and when there isn't? What I mean is, that while the data is being re-fetched, the old data is kept and can be accessed, semi-invalid as you call it. And in that time you can change it. But we know that this data after all is only cached data, I think it's ok we change it 100 times while it's semi invalid, and it will be reflected in the UI while refetching, the next time we subscribe to it, since it is mark as invalidated. I think that for the case where there aren't subscribers to the data and it is marked as invalid, we could keep the invalid data and if it bothers you, add an indicator, `isDataMarkedForInvalidation`, Obviously a better name than that but you get my point, and now the developer could make decisions based on that case.

What do you think?

**Comment by phryneas** [comment]
We can tell the difference when a component is using data and when it isn't. When it uses the hook, it "subscribes" to a cache value until it is either unmounted or subscribes to another cache value. That is our definition of "using a value". Actively telling the store "I need this" by using the query hook.

As for being "in line" with the other libraries:

The other two libraries don't even have a concept to invalidate data and automatically refetch it. Dealing with stale data is left 100% to the user.

If they had such a concept, they too would have to make some tradeoff at some point to make it practical - like we do here.  
This library already has quite a lot of concepts to understand and adding the concept of semi-stale data on top of it will probably do more harm than good.  

You are the first person ever coming up with this and there are easy ways to getting around it in your specific case.

Sorry, but I don't feel good about adding yet another concept on top. Things being the way they are now are a conscious design decision.

**Comment by omerman** [comment]
@phryneas I understand that it is a conscious design decision. and i read everything you wrote.

The only point I disagree is the one stating that other libraries dont have the concept of invalidating and automatically refetch it.

SWR has trigger method, which does exactly what I described. it marks the query's data as invalid, and if the data is currently observed, it performs a refetch. If not, it will refetch once the query is observed again. it will not remove the cached data until observed again & new data has arrived.

useQuery does the exact same thing, it has `invalidateQueries` api. it does that exactly.

I only suggested the 2 ideas because you told me that semi-invalid state is something RTK Query wants to avoid for bad UX purposes. To be perfectly honest, I think that if both of those libraries are ok with having invalid data that is stored until the next observation of the queries in question, you should be too. But offcourse thats a judgment call and not up to me.

**Comment by omerman** [comment]
@phryneas 
also regarding the "You are the first person ever coming up with this and there are easy ways to getting around it in your specific case."

I think I'm the first because this library is yet to be used broadly. 

and your simple solution is a hack that would need to be maintained regularly. it would not be a piece of code to write once and forget about it.

**Comment by phryneas** [comment]
> The only point I disagree is the one stating that other libraries dont have the concept of invalidating and automatically refetch it.

> SWR has the trigger method, which does exactly what I described. it marks the query's data as invalid, and if the data is currently observed, it performs a refetch. If not, it will refetch once the query is observed again. it will not remove the cached data until observed again & new data has arrived.

> useQuery does the exact same thing, it has invalidateQueries api. it does that exactly.

These are manual processes that have to be triggered by the user. You an at any point just manually trigger a query to be refetched in RTK-query as well, using a plethora of different mechanisms. If you do that instead of using `invalidates/provides`, no data will ever be thrown away, subscribed to or not.

For example: don't invalidate your list using the invalidation mechanism, but add an `onQueryStarted` lifecycle to your invalidating endpoint and just `dispatch(api.endpoints.yourList.initiate({ force: true, subscribe: false }))` from there. (pseudocode from the back of my head, might be different).
Or just do an optimistic update there and don't refetch at all. 

**My point is that they have no mechanism to invalidate data automatically, as we do with invalidates/provides.**

> Also the simple solution you supplied(thank you for that btw), is a hack that would need to be maintained on a regular basis. it would not be a piece of code to write once and forget about.

The same way you would have to manually maintain calling `trigger` or `invalidateQueries` with SWR or React-Query.

> I think I'm the first because this library is yet to be used broadly.

At this point it should be safe to say that we should have at least a thousand users or more (even during the RTKQ beta we had more than a hundred). If this were common, it would have come up more than just once by now.

**Comment by omerman** [comment]
@phryneas 

First of all, one of the main reasons I decided to move to RTK Query is because of the innovative tags usage.
I would probably keep using it even if you decide not to support the use case I brought up.

Last points I'll make and then feel free to close this issue if you are still not convinced:

 > For example: don't invalidate your list using the invalidation mechanism, but add an onQueryStarted lifecycle to your invalidating endpoint and just dispatch(api.endpoints.yourList.initiate({ force: true, subscribe: false })) from there. (pseudocode from the back of my head, might be different).
Or just do an optimistic update there and don't refetch at all.

First of all I want to make sure you understand, my important(in my opinion) enhancement suggestion,
I dont want to re-fetch after a mutation. As you mentioned before, If I refetch a query that is not being observed,
it would make it a potentially redundant request.

To clarify I suggested that invalidated query data will remain in the cache even if there are no observers to that query, because this behavior exists in similar libraries, and because I do not see it as bad UX, I see it as good UX, because what does it matter if redux keeps this data and refetches right away ONCE it is observed? I wouldn't even call it semi-invalid state, the data is cached, it's always potentially invalid, the data will go away on it's own once enough time is passed (keepUnusedDataFor)

To give you a simple & common use case that I'm sure 99% of the population would have:

Lets say I have Post entity & I have 2 queries, getAllPosts, and editPost.
Also 2 pages, AllPostsScreen, EditPostScreen.

I happen to scroll around the AllPostsScreen, and select a post to edit.
In the EditPostScreen I edit the post, and I invalidate the AllPosts query because it would contain some new data that I CANNOT modify myself in the client. Afterwards I'll forward the user to the AllPostsScreen.

On the use case above, RTK Query would chuck the old AllPosts query data because while on edit screen, there are no observers to the list query. Once I'm back in the AllPostsScreen I'll have to show the user a loader untill the list is refetched. In SWR/useQuery, the user will see the data he has just seen moments ago, with maybe some indication that his action is still somewhat in progress(a refetch is happening), and at some point the refetch will seamlessly display the most up to date data.

**Comment by phryneas** [comment]
Okay, let's just for a second assume we wanted to do this this:

Doing this is *not* removing those two lines.
It requires
* introducing a new `state` in the internal cache (this is kinda breaking, but internal)
* changes to the thunks
* changing the behavior of the selectors (this will *very likely* be breaking)
* introducing new behavior for the hooks (this will *very likely* be breaking)
* lots and *lots* and __*lots*__ of discussion and testing. You only have one chance to do something like this and get it right. As we have seen here, this already has cases where UX might become worse by introducing it and we would need to weigh all pros and cons.

So it will be a massive time sink and could not be released before we release a RTK 2.0.

Before we do a 2.0, there are still plans for a 1.7. 

Given my current work and personal situation, I cannot say when and how I will have enough time to even finish work on 1.7.  
We might even get a 1.6.2 with a few bug fixes before that, because I don't have much time for feature work at the moment.  
In addition, there is a lot of work to be done on the OpenAPI code generator, which at this point really should become a priority.

Also, there is already a lot of stuff on the Roadmap for 2.0. 
It might even not fit in here and only ship with a 3.0.

By the time we could ship it, it could even be irrelevant since much earlier (planned for 1.7) we will be shipping some kind of suspense-for-data-fetching-support, which will lead to a very similar experience to what you are suggesting. 
React will just defer rendering the component for the 50ms it takes you to fetch your list data - you will not get any loading spinners or empty components - and no stale data. It will just wait a few more frames and directly jump to fresh data. Which in most cases will be preferable over an stale-data-new-data jump.

So, I see that this could be a nice improvement. But that's it.  
It is not a glaring bug, it is working as currently intended (and edge cases like yours can be worked around) and there is so much stuff *more important* right now, that it could just be irrelevant by the time we get to it.  
And if by then it still is relevant (it could be, for Angular or Vue users), we will get to it then.

**Comment by omerman** [comment]
Understood. Thank you!
